### ISSUE TITLE
programmatically creating svgs

### ISSUE THREAD

**Issue by petermchale** [igvteam/igv.js#1138]
Given a list of loci, and a set of tracks, I'd like to *programmatically* create svgs for each locus, i.e., without manually inspecting each locus in a browser. How would you recommend doing that?

**Comment by petermchale** [comment]
This is a solution: https://github.com/brentp/jigv#automated-screenshots

But would like to know of other solutions.

**Comment by jrobinso** [comment]
@petermchale I'm not sure I see the connection between Brett's app referenced above and automatic svg creation.   This will be tricky to achieve with igv.js out of the box because loading data is asynchronous.  One could build an app around igv.js to do this but I'm not aware of any.   This would make a good project for someone, I'm sure there is an audience for it.

IGV desktop has the ability to run a batch script and create svgs (or pngs) programmatically.    As its a Java destkop app it must be run on a computer with a display, however others have used virtual display software to automate this on linux but I have no expertise in that.    A web search should turn up those solutions.   SVG from IGV desktop can be challenging to work with if you want to edit it, it includes nested artifacts of the Swing display system we can't control.

**Comment by petermchale** [comment]
Here is a solution, similar to Brent's above: 

https://gist.github.com/petermchale/3d8341b6929e64272046fa0a6eb5e659

It uses a workaround, `page.waitFor(10000)`, to "wait" for all data to load on the page before dumping a screenshot. 

Without  `page.waitFor(10000)`, a screenshot is created before all the data are loaded. I think this is because `igv.createBrowser` resolves before all data is loaded, as I've previously observed that 

`browser info: Fk {guid: "m7pp", namespace: ".browser_m7pp", config: {…}, parent: div#igv, $root: Bo.fn.init(1), …}`

appears in the console in Chrome's Dev Tools before all data appear in the webpage. 

Is there an event that is emitted by the page once all the data is loaded? If not, then can such an event be constructed? If the answer to either is "yes", then puppeteer can wait for that event before creating a screenshot, e.g., 

https://github.com/puppeteer/puppeteer/blob/main/examples/custom-event.js

**Comment by jrobinso** [comment]
No there's not, and this strategy will quickly become intractable if you use "search" to visit other regions.  You can try creating an empty browser (no tracks),  then add tracks with browser.loadTrack with a "sync" setting, as shown below.  loadTrack returns a promise to load a track, you can await it or use then.   This is not going to work if you "search" or do anything else that triggers another data load.   igv.js is not designed for this synchronous workflow.    @paul-shannon was trying to forces synchronous behavior for his igv-R project, not sure if he was ultimately successful or not.

```
            const trackConfig = {
                url: dataURL,
                indexURL: indexURL,
                sync: true
            }

           await browser.loadTrack(trackConfig)
```

**Comment by paul-shannon** [comment]
It worked great.  I execute my callbacks to R from completed promises.

**Comment by petermchale** [comment]
Thanks, @jrobinso . 

Awaiting the loading of a track seems to work in that I now see the track in the png but it is not a complete solution in that the promise seems to resolve before the spinner has disappeared (see spinner in attached png). Is there a way to await the spinner or an option to disable it? 

I don't see `sync` listed as an option at https://github.com/igvteam/igv.js/wiki/Tracks-2.0, and its value doesn't seem to affect the rendering in the final png. 

Also, the png does not show the reference genome (compare two attached pngs). Is there a way to await the loading of the genome too? 

![sample_vcf chr10-10,266,955-10,267,343](https://user-images.githubusercontent.com/19557137/88687129-1c428180-d0b5-11ea-8bb9-1aa293041c17.png)

![sample_vcf chr10-10,266,955-10,267,343 complete](https://user-images.githubusercontent.com/19557137/88687450-83603600-d0b5-11ea-8b7e-e0b0f8f0b839.png)

**Comment by jrobinso** [comment]
@petermchale I don't see your trackConfig in the above snippet.   That's where you set sync: true.

But as you point out this is a limited solution, only affecting loadTrack.   This will have no effect on the sequence.   I don't think there is a solution here unless you want to dig into the code,  however I can't really even encourage a pull request as the app is highly tuned to normal asynchronous usage.   Trying to synchronize everything would probably result in too many flags and switches to accept.  

We've had a solution for this with IGV desktop for a long time, and there are 3rd party tools written around it in python and R, maybe other languages.   A new solution based on igv.js would be cool, but is not something I can justify prioritizing with my limited time at the moment.

**Comment by petermchale** [comment]
OK. 

I've updated my gist above to reflect all the changes mentioned in this thread, for the benefit of others who may visit this issue. 

I'll also ping @turner , @paul-shannon and @brentp just for their information.

**Comment by jrobinso** [comment]
@petermchale I've been thinking about how to approach this a little bit but haven't come up with anything helpful.   I think you would need to "await" all xmlhttp requests which of course would kill web performance.   I think an event based approach would be intractable.  It might look simple until you get into the details.

**Comment by petermchale** [comment]
@jrobinso It just occurred to me that there must be something in igv.js that detects when data are rendered as it seems to show a "spinner" icon until the data are rendered. To the extent that this is true, then perhaps one can use, e.g., puppeteer to wait until the page no longer contains any spinners before capturing a screenshot...

**Comment by petermchale** [comment]
@jrobinso ^^^

**Comment by jrobinso** [comment]
@petermchale Yes, possibly, I don't have time to pursue this now, is there a specific question here?

One approach worth trying would be to loop through all trackViews on the "browser" object calling updateViews() on each, awaiting completion of each call,  then do a screenshot.     Something like

await Promise.all(browser.trackViews.map(tv => tv.updateViews())
